### PR TITLE
[Tests-Only] Restore files and folder from trashbin with same name

### PR DESCRIPTION
**Pull request by swoichha**
## Description

- Restore files from trashbin with same name as folder
- Restore folder from trashbin with same name as file

## Related Issue
https://github.com/owncloud/phoenix/issues/1723

## Motivation and Context

- To restore file from trashbin with same name as folder
- To restore folder from trashbin with same name as file

## How Has This Been Tested?
- :robot: 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

**Comment by PVince81** [comment]
@swoichha please always prefix the subject with "Tests" or something that makes it clear that the PR is about adding tests, not actually implementing the feature. Thanks!

File: tests/acceptance/features/webUITrashbin/trashbinRestore.feature [hunk 1]
```diff
@@ -115,3 +115,40 @@ Feature: Restore deleted files/folders
     And as "user1" file "simple-folder-renamed/file-to-delete-and-restore" should not exist
     #And as "user1" file "simple-folder-renamed/file-to-delete-and-restore" should exist
     And as "user1" file "simple-folder/file-to-delete-and-restore" should not exist
+
+  @issue-B
```
Comment (kiranparajuli589) [inline tests/acceptance/features/webUITrashbin/trashbinRestore.feature]:
```suggestion
  @issue-1723
```

Comment (individual-it) [inline tests/acceptance/features/webUITrashbin/trashbinRestore.feature]: what is the issue number? if its fixed, no need to include the number at all

File: tests/acceptance/features/webUITrashbin/trashbinRestore.feature [hunk 2]
```diff
@@ -115,3 +115,40 @@ Feature: Restore deleted files/folders
...
+  Scenario: Delete and restore file with same name as folder
```
Comment (individual-it) [inline tests/acceptance/features/webUITrashbin/trashbinRestore.feature]:
```suggestion
  Scenario: Delete and restore a file that has the same name like a deleted folder
```

File: tests/acceptance/features/webUITrashbin/trashbinRestore.feature [hunk 3]
```diff
@@ -115,3 +115,40 @@ Feature: Restore deleted files/folders
...
+    Given the user has created file "a1.txt"
```
Comment (individual-it) [inline tests/acceptance/features/webUITrashbin/trashbinRestore.feature]: you can use a file from the skeleton folder

File: tests/acceptance/features/webUITrashbin/trashbinRestore.feature [hunk 4]
```diff
@@ -115,3 +115,40 @@ Feature: Restore deleted files/folders
...
+    And the user has created folder "a1.txt"
+    And the following folders have been deleted by user "user1"
+      | name   |
+      | a1.txt |
+    When the user browses to the trashbin page
+    And the user restores file "a1.txt" from the trashbin using the webUI
+    Then the success message "a1.txt was restored successfully" should be displayed on the webUI
+    And file "a1.txt" should not be listed on the webUI
+    And the folder "a1.txt" should be listed on the webUI
+    When the user browses to the files page using the webUI
```
Comment (individual-it) [inline tests/acceptance/features/webUITrashbin/trashbinRestore.feature]: check that using the API

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 5]
```diff
@@ -40,6 +40,15 @@ module.exports = {
       return this.getFileRowSelectorByFileName(fileName) +
         this.elements.fileActionsButtonInFileRow.selector
     },
+    /**
+     * Get Selector for Folder Action expander
+     * @param folderName
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]:
```suggestion
     * @param {string} folderName
```

File: tests/acceptance/stepDefinitions/filesContext.js [hunk 6]
```diff
@@ -357,6 +364,14 @@ Then('file/folder {string} should not be listed on the webUI', async function (f
   )
 })
 
+Then('the folder {string} should not be listed on the webUI', async function (folder) {
+  client.pause(5000)
```
Comment (individual-it) [inline tests/acceptance/stepDefinitions/filesContext.js]: don't use hard-coded pause times

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 7]
```diff
@@ -51,7 +60,23 @@ module.exports = {
...
+      return this.initAjaxCounters()
+        .useXpath()
+        .waitForElementVisible(fileActionsBtnSelector)
+        .click(fileActionsBtnSelector)
+        .waitForElementVisible(btnSelector)
+        .click(btnSelector)
+        .useCss()
+    },
+    /**
+     * this method does find out itself if the folder-action burger has to be clicked or not
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: burger??

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 8]
```diff
@@ -51,7 +60,23 @@ module.exports = {
...
+     * @param {string}folderName
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]:
```suggestion
     * @param {string} folderName
```

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 9]
```diff
@@ -168,6 +193,20 @@ module.exports = {
         .waitForOutstandingAjaxCalls()
         .useCss()
     },
+    /**
+     *
+     * @param folderName
+     *
+     */
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: be more descriptive, add `return` tag and specify `params type` also

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 10]
```diff
@@ -404,6 +443,19 @@ module.exports = {
 
       return this
     },
+    /**
+     * Wait for A folderRow with foldername to be visible
+     *
+     * @param folderName
+     */
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: same as above!

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 11]
```diff
@@ -428,6 +480,12 @@ module.exports = {
 
       return { btnSelector, fileActionsBtnSelector }
     },
+    getFolderRowButtonSelectorsByFolderName: function (folderName, action) {
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: add docs

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 12]
```diff
@@ -453,6 +511,16 @@ module.exports = {
       const element = this.elements.fileRowByName
       return util.format(element.selector, xpathHelper.buildXpathLiteral(fileName))
     },
+    /**
+     *
+     * @param folderName
+     *
+     * @returns {string}
+     */
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: be more descriptive, add `return` tag and specify `params` type also

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 13]
```diff
@@ -461,6 +529,15 @@ module.exports = {
       return this.getFileRowSelectorByFileName(fileName) +
         this.elements.fileLinkInFileRow.selector
     },
+    /**
+     *
+     * @param folderName
+     */
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: same as above

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 14]
```diff
@@ -482,6 +559,27 @@ module.exports = {
         .useCss()
       return isListed
     },
+
+    /**
+     *
+     * @param folder
+     *
+     */
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: same as above!

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 15]
```diff
@@ -856,6 +954,9 @@ module.exports = {
     fileLinkInFileRow: {
       selector: '//span[contains(@class, "file-row-name")]'
     },
+    folderLinkInFolderRow: {
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: why is this added? is it used somewhere??

File: tests/acceptance/stepDefinitions/filesContext.js [hunk 16]
```diff
@@ -333,6 +333,13 @@ Then('file/folder {string} should be listed on the webUI', function (folder) {
     .waitForFileVisible(folder)
 })
 
+Then('the folder {string} should be listed on the webUI', function (folder) {
```
Comment (kiranparajuli589) [inline tests/acceptance/stepDefinitions/filesContext.js]:
```suggestion
Then('the folder {string} should be listed on the webUI', function (folderName) {
```

File: tests/acceptance/stepDefinitions/filesContext.js [hunk 17]
```diff
@@ -333,6 +333,13 @@ Then('file/folder {string} should be listed on the webUI', function (folder) {
...
+  return client
+    .page
+    .FilesPageElement.filesList()
+    .waitForFolderVisible(folder)
```
Comment (kiranparajuli589) [inline tests/acceptance/stepDefinitions/filesContext.js]:
```suggestion
    .waitForFolderVisible(folderName)
```

File: tests/acceptance/stepDefinitions/filesContext.js (hunk 6 above)
Comment (kiranparajuli589) [inline tests/acceptance/stepDefinitions/filesContext.js]:
```suggestion
```

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 18]
```diff
@@ -410,72 +419,97 @@ module.exports = {
...
-      return this.getFileRowSelectorByFileName(fileName) +
-        this.elements.fileLinkInFileRow.selector
+
+    getFileLinkSelectorByFileName: function (fileName, elementType = 'file') {
+      if (elementType === 'file') {
+        return this.getFileRowSelectorByFileName(fileName) +
+          this.elements.fileLinkInFileRow.selector
+      } else if (elementType === 'folder') {
+        fileName = '"' + fileName + '"'
+        return (util.format(this.elements.fileRowByName.selector, fileName) + '//div/span[@role="button"]')
```
Comment (individual-it) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: the xpath should go into the elements section

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 19]
```diff
@@ -410,72 +419,97 @@ module.exports = {
...
+        const linkSelector = this.getFileLinkSelectorByFileName(path, elementType)
+        return this
+          .useXpath()
+          .waitForElementVisible(rowSelector)
+          .getAttribute(linkSelector, 'innerText', function (result) {
+            this.assert.strictEqual(result.value.trim(), path, 'displayed file name not as expected')
+          })
+          .useCss
+      } else if (elementType === 'folder') {
+        const linkSelector = this.getFileLinkSelectorByFileName(path, elementType)
```
Comment (individual-it) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: duplication with line 427

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 20]
```diff
@@ -410,72 +419,97 @@ module.exports = {
...
+     */
+    waitForFileWithPathVisible: function (path, elementType = 'file') {
+      if (elementType === 'file') {
+        const rowSelector = this.getFileRowSelectorByFileName(path, elementType)
+        const linkSelector = this.getFileLinkSelectorByFileName(path, elementType)
+        return this
+          .useXpath()
+          .waitForElementVisible(rowSelector)
+          .getAttribute(linkSelector, 'innerText', function (result) {
+            this.assert.strictEqual(result.value.trim(), path, 'displayed file name not as expected')
```
Comment (individual-it) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: why is the waiting and the name checking only done for files and not folders?

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js (hunk 18 above)
Comment (swoichha) [reply tests/acceptance/pageObjects/FilesPageElement/filesList.js]: fixed

File: tests/acceptance/features/webUITrashbin/trashbinRestore.feature [hunk 21]
```diff
@@ -115,3 +115,39 @@ Feature: Restore deleted files/folders
     And as "user1" file "simple-folder-renamed/file-to-delete-and-restore" should not exist
     #And as "user1" file "simple-folder-renamed/file-to-delete-and-restore" should exist
     And as "user1" file "simple-folder/file-to-delete-and-restore" should not exist
+
+  @issue-1723
+  Scenario: Delete and restore a file that hs the same name like a deleted folder
```
Comment (phil-davis) [inline tests/acceptance/features/webUITrashbin/trashbinRestore.feature]:
```suggestion
  Scenario: Delete and restore a file that has the same name as a deleted folder
```

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 22]
```diff
@@ -91,12 +98,10 @@ module.exports = {
     deleteFile: async function (fileName) {
       await this.waitForFileVisible(fileName)
 
-      await this
+      return this
```
[the remark this answers is not in the document]
Comment (swoichha) [reply tests/acceptance/pageObjects/FilesPageElement/filesList.js]: fixed

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js (hunk 20 above)
Comment (swoichha) [reply tests/acceptance/pageObjects/FilesPageElement/filesList.js]: fixed

File: tests/acceptance/features/webUITrashbin/trashbinRestore.feature (hunk 21 above)
Comment (swoichha) [reply tests/acceptance/features/webUITrashbin/trashbinRestore.feature]: fixed

File: tests/acceptance/features/webUITrashbin/trashbinRestore.feature [hunk 23]
```diff
@@ -115,3 +115,39 @@ Feature: Restore deleted files/folders
...
+    When the user browses to the trashbin page
+    And the user restores file "lorem.txt" from the trashbin using the webUI
+    Then the success message "lorem.txt was restored successfully" should be displayed on the webUI
+    And file "lorem.txt" should not be listed on the webUI
+    And folder "lorem.txt" should be listed on the webUI
+    When the user browses to the files page using the webUI
+    Then file "lorem.txt" should be listed on the webUI
+    And folder "lorem.txt" should not be listed on the webUI
+
+  Scenario: Delete and restore a folder that hs the same name like a deleted file
```
Comment (kiranparajuli589) [inline tests/acceptance/features/webUITrashbin/trashbinRestore.feature]:
```suggestion
  Scenario: Delete and restore a folder that has the same name like a deleted file
```

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js (hunk 18 above)
Comment (swoichha) [reply tests/acceptance/pageObjects/FilesPageElement/filesList.js]: fixed

File: tests/acceptance/features/webUITrashbin/trashbinRestore.feature [hunk 24]
```diff
@@ -115,3 +115,39 @@ Feature: Restore deleted files/folders
     And as "user1" file "simple-folder-renamed/file-to-delete-and-restore" should not exist
     #And as "user1" file "simple-folder-renamed/file-to-delete-and-restore" should exist
     And as "user1" file "simple-folder/file-to-delete-and-restore" should not exist
+
+  @issue-1723
+  Scenario: Delete and restore a file that has the same name like a deleted folder
```
Comment (kiranparajuli589) [inline tests/acceptance/features/webUITrashbin/trashbinRestore.feature]:
```suggestion
  Scenario: Delete and restore a file that has same name with a deleted folder
```

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 25]
```diff
@@ -36,8 +36,17 @@ module.exports = {
...
      * @returns {string}
      */
-    getFileActionBtnSelector: function (fileName) {
-      return this.getFileRowSelectorByFileName(fileName) +
+    getFileActionBtnSelector: function (fileName, elementType = 'file') {
+      return this.getFileRowSelectorByFileName(fileName, elementType) +
+        this.elements.fileActionsButtonInFileRow.selector
+    },
+    /**
+     * Get Selector for Folder Action expander
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]:
```suggestion
     * Get selector for folder action expander
```

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 26]
```diff
@@ -158,13 +163,18 @@ module.exports = {
     },
     /**
      *
-     * @param {string} fileName
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: why change this??

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 27]
```diff
@@ -158,13 +163,18 @@ module.exports = {
...
+     * @param fileName
+     * @param elementType
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]:
```suggestion
     * @param {string} elementType
```

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 28]
```diff
@@ -158,13 +163,18 @@ module.exports = {
...
+      await this
+        .initAjaxCounters()
+        .waitForFileWithPathVisible(fileName, elementType)
+      await this
         .useXpath()
-        .performFileAction(fileName, FileAction.restore)
+        .performFileAction(fileName, FileAction.restore, elementType)
+      await this
         .waitForOutstandingAjaxCalls()
         .useCss()
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: try to chain them...

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 29]
```diff
@@ -410,72 +419,93 @@ module.exports = {
...
+      const linkSelector = this.getFileLinkSelectorByFileName(path, elementType)
+      const rowSelector = this.getFileRowSelectorByFileName(path, elementType)
       return this
         .useXpath()
         .waitForElementVisible(rowSelector)
         .getAttribute(linkSelector, 'innerText', function (result) {
           this.assert.strictEqual(result.value.trim(), path, 'displayed file name not as expected')
         })
-        .useCss()
+        .useCss
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: why this change??

File: tests/acceptance/stepDefinitions/filesContext.js [hunk 30]
```diff
@@ -326,11 +326,11 @@ Then(/there should be no files\/folders listed on the webUI/, function () {
   })
 })
 
-Then('file/folder {string} should be listed on the webUI', function (folder) {
+Then(/^(file|folder) "([^"]*)" should be listed on the webUI$/, async (elementType, folder) => {
```
Comment (kiranparajuli589) [inline tests/acceptance/stepDefinitions/filesContext.js]:
```suggestion
Then(/^(file|folder) "([^"]*)" should be listed on the webUI$/, (elementType, folder) => {
```

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 31]
```diff
@@ -159,12 +155,15 @@ module.exports = {
     /**
      *
      * @param {string} fileName
+     * @param {string}elementType
```
Comment (kiranparajuli589) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]:
```suggestion
     * @param {string} elementType
```

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 32]
```diff
@@ -422,60 +423,80 @@ module.exports = {
...
-      const fileActionsBtnSelector = this.getFileActionBtnSelector(fileName)
+    /**
+     *
+     * @param {string} fileName
+     * @param {string} action
+     * @param {string} elementType
+     * @returns {{btnSelector: string, fileActionsBtnSelector: string}}
+     */
+    getFileRowButtonSelectorsByFileName: function (fileName, action, elementType = 'file') {
+      const btnSelector = this.getActionSelector(action)
```
Comment (individual-it) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: where is the fileName gone?

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 33]
```diff
@@ -422,60 +423,80 @@ module.exports = {
...
+    getFileLinkSelectorByFileName: function (fileName, elementType = 'file') {
+      if (elementType === 'file') {
+        return this.getFileRowSelectorByFileName(fileName) +
+          this.elements.fileLinkInFileRow.selector
+      } else if (elementType === 'folder') {
+        fileName = '"' + fileName + '"'
+        return (util.format(this.elements.fileRowByName.selector, fileName) + this.elements.fileLinkInFileRow.selector)
+      } else {
+        return this
+      }
```
Comment (individual-it) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: that logic can go into `getFileRowSelectorByFileName()`

File: tests/acceptance/features/webUITrashbin/trashbinRestore.feature (hunk 4 above)
Comment (individual-it) [reply tests/acceptance/features/webUITrashbin/trashbinRestore.feature]: this is not done yet

File: tests/acceptance/stepDefinitions/filesContext.js [hunk 34]
```diff
@@ -350,8 +350,8 @@ Then('the last uploaded folder should be listed on the webUI', async function ()
   return client
 })
 
-Then('file/folder {string} should not be listed on the webUI', async function (folder) {
-  const state = await client.page.FilesPageElement.filesList().isElementListed(folder)
+Then(/^(file|folder) "([^"]*)" should not be listed on the webUI$/, async (elementType, folder) => {
```
Comment (kiranparajuli589) [inline tests/acceptance/stepDefinitions/filesContext.js]:
```suggestion
Then(/^(file|folder) ("(?:[^"\\]|\\.)*"|'(?:[^'\\]|\\.)*') should be listed on the webUI$/, (type, folder) => {
```
try this!!

File: tests/acceptance/stepDefinitions/filesContext.js [hunk 35]
```diff
@@ -326,11 +326,11 @@ Then(/there should be no files\/folders listed on the webUI/, function () {
   })
 })
 
-Then('file/folder {string} should be listed on the webUI', function (folder) {
+Then(/^(file|folder) ("([^"]*)"|'([^"]*)') should be listed on the webUI$/, (elementType, folder) => {
```
Comment (kiranparajuli589) [inline tests/acceptance/stepDefinitions/filesContext.js]:
```suggestion
Then(/^(file|folder) ("(?:[^"\\]|\\.)*"|'(?:[^'\\]|\\.)*') should be listed on the webUI$/, (elementType, folder) => {
```

Comment (individual-it) [reply tests/acceptance/stepDefinitions/filesContext.js]: why `\\`?

Comment (kiranparajuli589) [reply tests/acceptance/stepDefinitions/filesContext.js]: to get escapse sequences..

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js (hunk 32 above)
Comment (swoichha) [reply tests/acceptance/pageObjects/FilesPageElement/filesList.js]: fileName is still there. It is used while calling function getFileActionBtnSelector in line 435

Comment (individual-it) [reply tests/acceptance/pageObjects/FilesPageElement/filesList.js]: I see, the call to `getActionSelector()` was wrong before, that function only takes one argument

File: tests/acceptance/features/webUITrashbin/trashbinRestore.feature (hunk 4 above)
Comment (swoichha) [reply tests/acceptance/features/webUITrashbin/trashbinRestore.feature]: Couldn't check this using API as both the file and folder has same name and it couldn't separate whether file or folder is present in the trashbin. So the api result is not as expected and the test fails

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js (hunk 33 above)
Comment (swoichha) [reply tests/acceptance/pageObjects/FilesPageElement/filesList.js]: done

**Comment by phil-davis** [comment]
@individual-it @kiranparajuli589 please review again

**Comment by kiranparajuli589** [comment]
LGTM :+1:

File: tests/acceptance/stepDefinitions/filesContext.js [hunk 36]
```diff
@@ -371,6 +378,13 @@ Then('file/folder {string} should not be listed on the webUI', async function (f
   )
 })
 
+Then('{string} {string} should not be listed on the webUI', async (elementType, folder) => {
```
Comment (individual-it) [inline tests/acceptance/stepDefinitions/filesContext.js]: I think that will be too confusing to have a very similar step one with quotes around "folder" and one without.
Better to have separate step for file and folder
